stats/csv/collector.go

```diff
@@ -80,32 +74,43 @@ func New(fs afero.Fs, tags stats.TagSet, config Config) (*Collector, error) {
 	fname := config.FileName.String
 
 	if fname == "" || fname == "-" {
-		logfile := nopCloser{os.Stdout}
 		return &Collector{
-			outfile:      logfile,
 			fname:        "-",
 			resTags:      resTags,
 			ignoredTags:  ignoredTags,
-			csvWriter:    csv.NewWriter(logfile),
+			csvWriter:    csv.NewWriter(os.Stdout),
 			row:          make([]string, 3+len(resTags)+1),
 			saveInterval: saveInterval,
+			closeFn:      func() error { return nil },
 		}, nil
 	}
 
-	logfile, err := fs.Create(fname)
+	logFile, err := fs.Create(fname)
 	if err != nil {
 		return nil, err
 	}
 
-	return &Collector{
-		outfile:      logfile,
+	c := Collector{
 		fname:        fname,
 		resTags:      resTags,
 		ignoredTags:  ignoredTags,
-		csvWriter:    csv.NewWriter(logfile),
 		row:          make([]string, 3+len(resTags)+1),
 		saveInterval: saveInterval,
-	}, nil
+	}
+
+	if strings.HasSuffix(fname, ".gz") {
+		outfile := gzip.NewWriter(logFile)
+		c.csvWriter = csv.NewWriter(outfile)
```

Comment:
I think you'll need to add an additional `Flush()` call to the `gzip.Writer`, as the CSV one will not propagate to it.

The current behavior is that the `.gz` file remains at 10 bytes (presumably the gzip header), and then the contents are written at once when the test is done. It would be better if this worked the same way as the non-compressed output where rows are written as soon as processed, to avoid keeping that data in memory and so the user can stream it.

Defining a `Collector.flushFn` closure here that does both calls should do the trick.

Reply:
Oops, thanks for the catch. Fixing this right away!

Reply:
Sorry, that I am late, but I think ... this is not needed. 

@imiric did you test with a very small amount of metrics? Because if so ... chances are that they just fit in the default `bufio.Writer` buffer and would've been written to disk if there was more data. 

Calling Flush constantly, while probably will be a noop in most cases, doesn't seem like a requirement. I don't do it in the `json` output and it did output data to the disk ;)

Reply:
Yeah, my test was short, so that's likely what happened. I just tested again on `c7de12e4` with a larger test and the .gz file was written to after about ~20kb.

Apologies @thejasbabu, I think we can revert the `flushFn` changes. The immediate flushing is not that important.